dali/operators/reader/loader/video/frames_decoder_base.cc

```diff
@@ -651,21 +612,68 @@ bool FramesDecoderBase::ReadRegularFrame(uint8_t *data) {
   return false;
 }
 
+bool FramesDecoderBase::AvSeekFrame(int64_t timestamp, int frame_id) {
+  if (is_seekable_.has_value() && !is_seekable_.value()) {
+    LOG_LINE << "Not seekable, returning directly" << std::endl;
+    return false;
+  }
+
+  if (av_state_->codec_) {
+    avcodec_flush_buffers(av_state_->codec_ctx_);
+  }
+
+  int ret = av_seek_frame(av_state_->ctx_, av_state_->stream_id_, timestamp, AVSEEK_FLAG_FRAME);
+  if (!is_seekable_.has_value()) {
+    is_seekable_ = ret >= 0;
+  }
+  if (ret < 0) {
+    return false;
+  } else {
+    next_frame_idx_ = frame_id;
+    return true;
+  }
+}
+
 void FramesDecoderBase::Reset() {
-  LOG_LINE << "Resetting decoder" << std::endl;
+  if (AvSeekFrame(0, 0)) {
+    LOG_LINE << "Reset: Seeked to first frame." << std::endl;
+  } else {
+    LOG_LINE << "Reset: Failed to seek to first frame. Reopening stream." << std::endl;
+  }
 
-  next_frame_idx_ = 0;
+  bool require_available_avcodec = av_state_->codec_ != nullptr;
+  int stream_id = av_state_->stream_id_;
+  av_state_ = std::make_unique<AvState>();
+  DALI_ENFORCE((av_state_->ctx_ = avformat_alloc_context()), "Could not alloc avformat context");
 
-  if (flush_state_) {
-    flush_state_ = false;
+  int ret = -1;
+  if (memory_video_file_.has_value()) {
+    memory_video_file_.value().Seek(0, SEEK_SET);
+
+    uint8_t *av_io_buffer = static_cast<uint8_t *>(av_malloc(default_av_buffer_size));
+    AVIOContext *av_io_context = avio_alloc_context(
+        av_io_buffer, default_av_buffer_size, 0, &memory_video_file_.value(),
+        detail::read_memory_video_file, nullptr, detail::seek_memory_video_file);
+    av_state_->ctx_->pb = av_io_context;
```

Comment:
1) What if allocation fails?
2) Shouldn't we clean up `av_state_->ctx_->pb`?
```
 auto custom_ctx = ctx_->pb;
      auto custom_buffer = ctx_->pb ? ctx_->pb->buffer : nullptr;
      if (custom_dealloc) {
        av_freep(&custom_buffer);
        avio_context_free(&custom_ctx);
      }
```

Reply:
1. I'll handle that
2. No need, it already gets cleaned up when we do `av_state_ = std::make_unique<AvState>();`